### PR TITLE
[Enhancement] str_to_date supports incomplete format types in FE

### LINKED ISSUE #32625: Enhance str_to_date() generate execution plans in FE
## Enhancement
now it is only support str_to_date('2023-01-01','%Y-%m-%d') 和str_to_date('2023-01-01 00:00:00','%Y-%m-%d %H:%i:%s') when FE generate execution plans . if  str_to_date('2023-01-01 00:00:00','%Y-%m-%d') or str_to_date('2023-01-01','%Y-%m-%d %H:%i:%s') in where clause and compare with partition filed, it will scan all table but not perform partition cropping
截屏2023-10-12 15 28 43
截屏2023-10-12 15 29 18

### PR DESCRIPTION
**Pull request by even986025158**
str_to_date supports incomplete format types in FE

Fixes #32625

## What type of PR is this:

- [ ] BugFix
- [ ] Feature
- [x] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

Does this PR entail a change in behavior?

- [x] Yes, this PR will result in a change in behavior.
- [ ] No, this PR will not result in a change in behavior.

If yes, please specify the type of change:

- [ ] Interface/UI changes: syntax, type conversion, expression evaluation, display information
- [ ] Parameter changes: default values, similar parameters but with different default values
- [x] Policy changes: use new policy to replace old one, functionality automatically enabled
- [ ] Feature removed
- [ ] Miscellaneous: upgrade & downgrade compatibility, etc.

截屏2023-10-12 16 59 00

## Checklist:

- [x] I have added test cases for my bug fix or my new feature
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function

## Bugfix cherry-pick branch check:

- [x] I have checked the version labels which the pr will be auto-backported to the target branch
  - [x] 3.2
  - [ ] 3.1
  - [ ] 3.0
  - [ ] 2.5

### CODE REVIEW

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]


:heart_eyes: **pass** : 0 / 0 (0%)

File: fe/fe-core/src/test/java/com/starrocks/sql/plan/ExpressionTest.java [hunk 1]
```diff
@@ -376,20 +376,20 @@ public void testNullAddNull() throws Exception {
...
         starRocksAssert.query(sql)
-                .explainContains("nullif('2021-01-12 00:00:00', str_to_date('2020-11-02', '%Y-%m-%d %H:%i:%s'))");
+                .explainContains("nullif('2021-01-12 00:00:00', '2020-11-02 00:00:00'");
 
         sql = "select if(3, 4, 5);";
         starRocksAssert.query(sql).explainContains("if(CAST(3 AS BOOLEAN), 4, 5)");
 
         sql = "select ifnull(date('2021-01-12'), 123);";
-        starRocksAssert.query(sql).explainContains("20210112");
+        starRocksAssert.query(sql).explainContains("CAST('2021-01-12' AS INT)");
```
Comment (ZiheLiu) [inline fe/fe-core/src/test/java/com/starrocks/sql/plan/ExpressionTest.java:392]: This UT failed. 
```suggestion
        starRocksAssert.query(sql).explainContains("20210112");
```

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rewrite/ScalarOperatorFunctions.java [hunk 2]
```diff
@@ -317,12 +319,43 @@ public static ConstantOperator dateParse(ConstantOperator date, ConstantOperator
...
-            LocalDate ld = LocalDate.from(builder.withResolverStyle(ResolverStyle.STRICT).parse(dateStr));
+            LocalDate ld;
+            try {
+                ld = LocalDate.from(builder.withResolverStyle(ResolverStyle.STRICT).parse(dateStr));
+            } catch (DateTimeParseException e) {
+                // eg. str_to_date('2022-10-18 00:00:00','%Y-%m-%d');
+                if (dateStr.length() > e.getErrorIndex()) {
+                    ld = LocalDate.from(builder.withResolverStyle(ResolverStyle.STRICT)
+                            .parse(dateStr.substring(0, e.getErrorIndex())));
+                } else {   // eg. str_to_date('2022-10','%Y-%m-%d');
```
Comment (ZiheLiu) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rewrite/ScalarOperatorFunctions.java:348]: `str_to_date('2022-10','%Y-%m-%d')` returns `null` in BE, so we needn't this `else` branch.

```suggestion
                ld = LocalDate.from(builder.withResolverStyle(ResolverStyle.STRICT)
                            .parse(dateStr.substring(0, e.getErrorIndex())));
```

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rewrite/ScalarOperatorFunctions.java [hunk 3]
```diff
@@ -317,12 +319,43 @@ public static ConstantOperator dateParse(ConstantOperator date, ConstantOperator
...
+                } else {   // eg. str_to_date('2022-10-18 00:00','%Y-%m-%d %H:%s:%i');
+                    int length = dateStr.length();
+                    //Since the length of the actual format cannot be obtained, it needs to be filled first. If it fails again,
+                    //the length of the actual format can be obtained, and the format can be converted.
+                    try {
+                        ldt = LocalDateTime.from(builder.withResolverStyle(ResolverStyle.STRICT)
+                                .parse(dateStr + DEFAULT_DATE_TIME.substring(length)));
+                    } catch (DateTimeParseException pe) {
+                        ldt = LocalDateTime.from(builder.withResolverStyle(ResolverStyle.STRICT)
+                                .parse(dateStr + DEFAULT_DATE_TIME.substring(length, pe.getErrorIndex())));
```
Comment (ZiheLiu) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rewrite/ScalarOperatorFunctions.java:334]: As for `str_to_date(str, pattern)`, if `pattern` is longer than `str` and `str` is a complete date but not **datetime**, this conversion can be successful. We should check whether it is a valid date format here, that is, `dateStr` must contain the complete `%Y-%m-%d` part.
For example, 
- `str_to_date('2022-10-18 00:00','%Y-%m-%d %H:%s:%i')` returns `2022-10-18 00:00:00`.
- `str_to_date('2022-10','%Y-%m-%d %H:%s:%i')` returns `null`.

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rewrite/ScalarOperatorFunctions.java (hunk 2 above)
Comment (LiShuMing) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rewrite/ScalarOperatorFunctions.java:348]: You can use `set disable_function_fold_constants=off/on` to check whether FE/BE's behavior is the same or not.

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rewrite/ScalarOperatorFunctions.java (hunk 3 above)
Comment (even986025158) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rewrite/ScalarOperatorFunctions.java]: I has changed it by BE code, please help to review it, thanks!

**Pull request closed without merging**